Add item requests Sprint 15

V-Levchenkov opened the pull request.

V-Levchenkov committed:
Sprint 15( add-item-requests) добавлен функционал.

V-Levchenkov committed:
Sprint 15( add-item-requests) add tests

kukolev commented on src/main/java/ru/practicum/shareit/booking/dto/BookingMapper.java:
```diff
@@ -35,6 +35,13 @@ public Booking fromSimpleToBooking(BookingDtoSimple bookingDtoSimple) {
                 Status.WAITING);
     }
 
+    public BookingDtoSimple toBookingDtoSimple(Booking booking) {
```
Метод нужно сделать приватным, т.к. он является вспомогательными и не нужно, чтобы к нему был доступ извне.

V-Levchenkov replied:
Не согласен, т.к. это Dto и не несет никакой угрозs безопасности, и дабы не плодить объекты в коде считаю, что лучше оставить так.

kukolev replied:
На самом деле дела не в приватности. Написал в новом замечании. Если хотите, то давай обсудим в слаке.

kukolev commented on src/main/java/ru/practicum/shareit/booking/repository/BookingRepository.java:
```diff
@@ -10,40 +11,50 @@
+    List<Booking> searchBookingByItemOwnerId(long id, Pageable pageable);
 
-    List<Booking> searchBookingByBookerIdAndItemIdAndEndIsBefore(long id, long itemId, LocalDateTime time);
+    List<Booking> searchBookingByBookerIdAndItemIdAndEndIsBeforeAndStatus(long id, long itemId,
```
По моему такие методы очень плохо читаются в коде. Будет лучше, если названия будут короткими, а логика поиска будет реализована в `@Query`

V-Levchenkov replied:
В моем видении так будет более информативно, т.к. понятно чего хочет этот метод, так же как и в прошлом спринте мы уже решили, что такая реализация будет соответствовать представленной теоретической части.

kukolev replied:
Еще раз повторю, что в продакшене так никто не делает. Это плохая практика и делать так я не советую, т.к. вы привыкаете к плохой практике. Окей, дело ваше, т.к. строгих стандартов на длину названий методов в проекте нет, но мои рекомендации в силе.

kukolev commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -16,11 +20,11 @@
 import java.util.stream.Collectors;
 
 @Service
+@Slf4j
```
Логику сервисов лучше покрыть логами. Так будет проще анализировать поведение программы, если что-то пойдет не так.

V-Levchenkov replied:
done

kukolev commented on src/main/java/ru/practicum/shareit/requests/service/ItemRequestServiceImpl.java:
```diff
@@ -0,0 +1,78 @@
+import java.util.stream.Collectors;
+
+@Service
+@Slf4j
```
Логику сервисов лучше покрыть логами. Так будет проще анализировать поведение программы, если что-то пойдет не так.

V-Levchenkov replied:
done

kukolev commented on src/test/java/ru/practicum/shareit/booking/repository/BookingRepositoryTest.java:
```diff
@@ -0,0 +1,243 @@
+class BookingRepositoryTest {
+
+    @Autowired
+    ItemRepository itemRepository;
```
Нарушен принцип инкапсуляции. Поля класса не должны быть видимы другим классам. Лучше сделать поле приватным и сделать геттер и сеттер для полей, если нужен доступ извне.

V-Levchenkov replied:
done

kukolev commented on src/test/java/ru/practicum/shareit/booking/repository/BookingRepositoryTest.java:
```diff
@@ -0,0 +1,243 @@
+    }
+
+    @Test
+    void searchBookingByItem_Owner_IdAndStartIsAfter() {
```
Этот тест лучше разбить на несколько небольших тестов, каждый из которых проверяет что-то конкретное. Так будет проще понять что пошло не так, если один или несколько тестов упадут.

V-Levchenkov replied:
done

kukolev commented on src/test/java/ru/practicum/shareit/booking/repository/BookingRepositoryTest.java:
```diff
@@ -0,0 +1,243 @@
+    }
+
+    @Test
+    void findByBooker_IdAndStartAfter() {
```
Этот тест лучше разбить на несколько небольших тестов, каждый из которых проверяет что-то конкретное. Так будет проще понять что пошло не так, если один или несколько тестов упадут.

V-Levchenkov replied:
done

kukolev commented on src/test/java/ru/practicum/shareit/booking/repository/BookingRepositoryTest.java:
```diff
@@ -0,0 +1,243 @@
+    }
+
+    @Test
+    void findBookingsByItem_IdAndStartIsAfterOrderByStartDesc() {
```
Этот тест лучше разбить на несколько небольших тестов, каждый из которых проверяет что-то конкретное. Так будет проще понять что пошло не так, если один или несколько тестов упадут.

V-Levchenkov replied:
done

kukolev commented on src/test/java/ru/practicum/shareit/booking/repository/BookingRepositoryTest.java:
```diff
@@ -0,0 +1,243 @@
+    }
+
+    @Test
+    void findCurrentBookingsByBooker_Id() {
```
Этот тест лучше разбить на несколько небольших тестов, каждый из которых проверяет что-то конкретное. Так будет проще понять что пошло не так, если один или несколько тестов упадут.

V-Levchenkov replied:
refactor, done

kukolev commented on src/test/java/ru/practicum/shareit/booking/repository/BookingRepositoryTest.java:
```diff
@@ -0,0 +1,243 @@
+    }
+
+    @Test
+    void findCurrentBookingsByItem_Owner_Id() {
```
Этот тест лучше разбить на несколько небольших тестов, каждый из которых проверяет что-то конкретное. Так будет проще понять что пошло не так, если один или несколько тестов упадут.

V-Levchenkov replied:
refactor, done

kukolev commented on src/test/java/ru/practicum/shareit/booking/service/BookingServiceImplTest.java:
```diff
@@ -0,0 +1,251 @@
+    }
+
+    @Test
+    void findAll() {
```
Этот тест лучше разбить на несколько небольших тестов, каждый из которых проверяет что-то конкретное. Так будет проще понять что пошло не так, если один или несколько тестов упадут.

V-Levchenkov replied:
refactor, done

kukolev commented on src/test/java/ru/practicum/shareit/booking/service/BookingServiceImplTest.java:
```diff
@@ -0,0 +1,251 @@
+    }
+
+    @Test
+    void save() {
```
Этот тест лучше разбить на несколько небольших тестов, каждый из которых проверяет что-то конкретное. Так будет проще понять что пошло не так, если один или несколько тестов упадут.

V-Levchenkov replied:
done

kukolev commented on src/test/java/ru/practicum/shareit/booking/service/BookingServiceImplTest.java:
```diff
@@ -0,0 +1,251 @@
+import static org.mockito.Mockito.*;
+import static ru.practicum.shareit.booking.enums.Status.*;
+
+class BookingServiceImplTest {
```
Названия тестов в классе не отображают предназначение теста. Названия тестов можно делать многословными, чтобы из названия было понятно что и как проверяет тест.

V-Levchenkov replied:
done

kukolev commented on src/test/java/ru/practicum/shareit/booking/service/BookingServiceImplTest.java:
```diff
@@ -0,0 +1,251 @@
+    }
+
+    @Test
+    void findAllByItemOwnerId() {
```
Этот тест лучше разбить на несколько небольших тестов, каждый из которых проверяет что-то конкретное. Так будет проще понять что пошло не так, если один или несколько тестов упадут.

V-Levchenkov replied:
refactor, done

kukolev commented on src/test/java/ru/practicum/shareit/item/controller/ItemControllerTest.java:
```diff
@@ -0,0 +1,179 @@
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@ExtendWith(MockitoExtension.class)
+class ItemControllerTest {
```
Названия тестов в классе не отображают предназначение теста. Названия тестов можно делать многословными, чтобы из названия было понятно что и как проверяет тест.

V-Levchenkov replied:
done

kukolev commented on src/test/java/ru/practicum/shareit/item/repository/CommentRepositoryTest.java:
```diff
@@ -0,0 +1,54 @@
+import static org.junit.jupiter.api.Assertions.*;
+
+@DataJpaTest
+class CommentRepositoryTest {
```
Названия тестов в классе не отображают предназначение теста. Названия тестов можно делать многословными, чтобы из названия было понятно что и как проверяет тест.

V-Levchenkov replied:
done

kukolev commented on src/test/java/ru/practicum/shareit/item/repository/CommentRepositoryTest.java:
```diff
@@ -0,0 +1,54 @@
+class CommentRepositoryTest {
+
+    @Autowired
+    ItemRepository itemRepository;
```
Нарушен принцип инкапсуляции. Поля класса не должны быть видимы другим классам. Лучше сделать поле приватным и сделать геттер и сеттер для полей, если нужен доступ извне.

V-Levchenkov replied:
done

kukolev commented on src/test/java/ru/practicum/shareit/item/repository/ItemRepositoryTest.java:
```diff
@@ -0,0 +1,81 @@
+class ItemRepositoryTest {
+
+    @Autowired
+    ItemRepository itemRepository;
```
Нарушен принцип инкапсуляции. Поля класса не должны быть видимы другим классам. Лучше сделать поле приватным и сделать геттер и сеттер для полей, если нужен доступ извне.

V-Levchenkov replied:
done

kukolev commented on src/test/java/ru/practicum/shareit/item/service/ItemServiceImplTest.java:
```diff
@@ -0,0 +1,203 @@
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.Mockito.*;
+
+class ItemServiceImplTest {
```
Названия тестов в классе не отображают предназначение теста. Названия тестов можно делать многословными, чтобы из названия было понятно что и как проверяет тест.

V-Levchenkov replied:
done

kukolev commented on src/test/java/ru/practicum/shareit/item/service/ItemServiceImplTest.java:
```diff
@@ -0,0 +1,203 @@
+    }
+
+    @Test
+    void saveComment() {
```
Этот тест лучше разбить на несколько небольших тестов, каждый из которых проверяет что-то конкретное. Так будет проще понять что пошло не так, если один или несколько тестов упадут.

V-Levchenkov replied:
done

kukolev commented on src/test/java/ru/practicum/shareit/requests/controller/ItemRequestControllerTest.java:
```diff
@@ -0,0 +1,132 @@
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@ExtendWith(MockitoExtension.class)
+class ItemRequestControllerTest {
```
Названия тестов в классе не отображают предназначение теста. Названия тестов можно делать многословными, чтобы из названия было понятно что и как проверяет тест.

V-Levchenkov replied:
done

kukolev commented on src/test/java/ru/practicum/shareit/requests/repository/ItemRequestRepositoryTest.java:
```diff
@@ -0,0 +1,67 @@
+class ItemRequestRepositoryTest {
+
+    @Autowired
+    ItemRepository itemRepository;
```
Нарушен принцип инкапсуляции. Поля класса не должны быть видимы другим классам. Лучше сделать поле приватным и сделать геттер и сеттер для полей, если нужен доступ извне.

V-Levchenkov replied:
done

kukolev commented on src/test/java/ru/practicum/shareit/requests/repository/ItemRequestRepositoryTest.java:
```diff
@@ -0,0 +1,67 @@
+import static org.junit.jupiter.api.Assertions.*;
+
+@DataJpaTest
+class ItemRequestRepositoryTest {
```
Названия тестов в классе не отображают предназначение теста. Названия тестов можно делать многословными, чтобы из названия было понятно что и как проверяет тест.

V-Levchenkov replied:
Done

kukolev commented on src/test/java/ru/practicum/shareit/requests/service/ItemRequestServiceImplTest.java:
```diff
@@ -0,0 +1,118 @@
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.Mockito.*;
+
+class ItemRequestServiceImplTest {
```
Названия тестов в классе не отображают предназначение теста. Названия тестов можно делать многословными, чтобы из названия было понятно что и как проверяет тест.

V-Levchenkov replied:
done

kukolev commented on src/test/java/ru/practicum/shareit/user/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,96 @@
+class UserControllerTest {
+
+    @MockBean
+    UserService userService;
```
Нарушен принцип инкапсуляции. Поля класса не должны быть видимы другим классам. Лучше сделать поле приватным и сделать геттер и сеттер для полей, если нужен доступ извне.

V-Levchenkov replied:
done

kukolev commented on src/test/java/ru/practicum/shareit/user/service/UserServiceImplTest.java:
```diff
@@ -0,0 +1,102 @@
+
+class UserServiceImplTest {
+
+    UserService userService;
```
Нарушен принцип инкапсуляции. Поля класса не должны быть видимы другим классам. Лучше сделать поле приватным и сделать геттер и сеттер для полей, если нужен доступ извне.

V-Levchenkov replied:
done

kukolev commented on src/test/java/ru/practicum/shareit/user/service/UserServiceImplTest.java:
```diff
@@ -0,0 +1,102 @@
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.Mockito.*;
+
+class UserServiceImplTest {
```
Названия тестов в классе не отображают предназначение теста. Названия тестов можно делать многословными, чтобы из названия было понятно что и как проверяет тест.

V-Levchenkov replied:
done

kukolev commented on src/test/java/ru/practicum/shareit/user/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,96 @@
+
+@WebMvcTest(UserController.class)
+@AutoConfigureMockMvc
+class UserControllerTest {
```
Названия тестов в классе не отображают предназначение теста. Названия тестов можно делать многословными, чтобы из названия было понятно что и как проверяет тест.

V-Levchenkov replied:
done

V-Levchenkov committed:
Первые исправления по ревью

V-Levchenkov committed:
Второй подход исправлений по ревью, рефактор тестов

V-Levchenkov committed:
Второй подход исправлений по ревью, рефактор тестов

V-Levchenkov committed:
Второй подход исправлений по ревью, рефактор тестов

V-Levchenkov committed:
Второй подход исправлений по ревью, рефактор тестов

V-Levchenkov committed:
рефактор тестов,устранение замечаний.

V-Levchenkov committed:
checkStyle

kukolev commented on src/main/java/ru/practicum/shareit/booking/dto/BookingMapper.java:
```diff
@@ -35,6 +35,13 @@ public Booking fromSimpleToBooking(BookingDtoSimple bookingDtoSimple) {
                 Status.WAITING);
     }
 
+    public BookingDtoSimple toBookingDtoSimple(Booking booking) {
```
В предыдущем замечании я неверно идентифицировал проблему. Проблема серьезней. Судя по всему этот код нужен только для вспомогательных целей в тестах.

Важно, чтобы тестовый код не перемешивался с основным. Иначе есть риск, что тестовый код попадет в продакшен.  Этот маппинг и сам класс BookingDtoSimple нужно перенести в тесты. В тестах можно сделать отдельную тестовую утилиту в которой будет этот маппинг, это распространенная практика.

V-Levchenkov replied:
done

kukolev commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -39,58 +43,65 @@ public BookingServiceImpl(BookingRepository bookingRepository, ItemRepository it
 
     @Override
     public BookingDto findById(long bookingId, long userId) {
+        log.info("Поиск по bookingId");
```
При логировании лучше добавлять информацию о текущих значениях параметров, чтобы при анализе было понятно о чем идет речь. Здесь будет полезно добавить значение переменной «bookingId».

Аналогично рекомендую поправить в остальных частях кода.

V-Levchenkov replied:
done + userSevice logs

V-Levchenkov committed:
Ревью 2:
Перенос метода для теста в Utill
Обновление логирования согласно ревью + добавление недостающего логирования в User

kukolev commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -43,12 +43,12 @@ public BookingServiceImpl(BookingRepository bookingRepository, ItemRepository it
     @Override
     public BookingDto findById(long bookingId, long userId) {
-        log.info("Поиск по bookingId");
+        log.info("Поиск по bookingId: " + bookingId);
```
При логировании лучше использовать возможности библиотеки и пользоваться плейсхолдерами `{}` вместо конкатенации строк. Лучше так: `log.warn("Some warning text for {}", val)` вместо `log.warn("Some warning text for " + val)` или `log.warn(String.format("Some warning text for %s", val))` . Так код смотрится проще и тратится меньше ресурсов.

Аналогично нужно поправить в остальных местах кода, где есть логирование.

V-Levchenkov replied:
done

kukolev commented on src/test/java/ru/practicum/shareit/booking/bookingTestUtils/TestUtils.java:
```diff
@@ -0,0 +1,13 @@
+import ru.practicum.shareit.booking.model.Booking;
+
+public class TestUtils {
+    public BookingDtoSimple toBookingDtoSimple(Booking booking) {
```
Лучше сделать метод статическим, т.к. в экземпляре класса нет смысла и тогда не нужно будет создавать экземпляры класса в тестах.

V-Levchenkov replied:
done

V-Levchenkov committed:
исправление замечаний по 3-му ревью

V-Levchenkov committed:
правки под тест

V-Levchenkov committed:
правки под тест
